source/librii/rarc/RARC.cpp

```diff
@@ -0,0 +1,1281 @@
+    return std::unexpected(
+        "Saved data mismatched the recorded data size provided!");
+
+  return result;
+}
+
+Result<void> RecalculateArchiveIDs(ResourceArchive& arc) {
+  std::vector<int> parent_stack = {
+      -1,
+  };
+  std::vector<int> walk_stack = {};
+  int folder_id = 0;
+
+  for (auto node = arc.nodes.begin(); node != arc.nodes.end(); node++) {
+    for (auto walk = walk_stack.begin(); walk != walk_stack.end();) {
+      if ((*walk) == std::distance(arc.nodes.begin(), node)) {
+        walk = walk_stack.erase(walk);
+        parent_stack.pop_back();
+      } else {
+        walk++;
+      }
+    }
+    if (node->is_folder()) {
+      node->id = folder_id++;
+      node->folder.parent = parent_stack.back();
+      walk_stack.push_back(node->folder.sibling_next);
+      parent_stack.push_back(node->id);
+    }
+  }
+
+  children_info_type throwaway;
+  const std::vector<ResourceArchive::Node>& calc_list =
+      TRY(rarcProcessNodesForSave(arc, throwaway));
+
+  s16 file_id = -1; // Adjust for root node
+  for (auto& node : calc_list) {
+    if (node.is_folder()) {
+      file_id++;
+      continue;
+    }
+
+    for (auto& target : arc.nodes) {
+      if (target.id == node.id && target.name == node.name) {
+        target.id = file_id++;
+        break;
+      }
+    }
+  }
+
+  return {};
+}
+
+Result<void> ExtractResourceArchive(const ResourceArchive& arc,
+                                    std::filesystem::path out) {
+  auto tmp = out;
+  std::vector<u32> stack;
+  u32 i = 0;
+  for (auto& node : arc.nodes) {
+    if (!stack.empty()) {
+      while (stack.back() == i) {
+        stack.resize(stack.size() - 1);
+        tmp = tmp.parent_path();
+      }
+    }
+    ++i;
+    if (node.is_folder()) {
+      stack.push_back(node.folder.sibling_next);
+      tmp /= node.name;
+      std::filesystem::create_directory(tmp);
+    } else {
+      auto fpath = (tmp / node.name).string();
+      std::ofstream fout(fpath, std::ios::binary | std::ios::ate);
+      fout.write(reinterpret_cast<const char*>(node.data.data()),
+                 node.data.size());
+    }
+  }
+  return {};
+}
+
+static void
+get_sorted_directory_list_r(const std::filesystem::path& path,
+                            std::vector<std::filesystem::path>& out) {
+  std::vector<std::filesystem::path> dirs;
+  for (auto&& it : std::filesystem::directory_iterator{path}) {
+    if (it.is_directory())
+      dirs.push_back(it.path());
+    else
+      out.push_back(it.path());
+  }
+  for (auto& dir : dirs) {
+    out.push_back(dir);
+    get_sorted_directory_list_r(dir, out);
+  }
+}
+
+Result<ResourceArchive> CreateResourceArchive(std::filesystem::path root) {
+  struct Entry {
+    std::string str;
+    bool is_folder = false;
+
+    std::vector<u8> data;
+    std::string name;
+
+    int depth = -1;
+    int parent = 0; // Default files put parent as 0 for root
+    int siblingNext = -1;
+
+    bool operator<(const Entry& rhs) const {
```

Comment:
This isn't a true partial ordering since it violates `(A < B  && B < A) => A != B`. I'd rather this be pulled out as a lambda "rarc_sort_order" or something.

Reply:
Turns out this went unused anyway (was probably planning to use std::sort at some point).